queue/retry.go

```diff
@@ -9,11 +9,42 @@ import (
 	"github.com/evergreen-ci/utility"
 	"github.com/mongodb/amboy"
 	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/level"
 	"github.com/mongodb/grip/message"
 	"github.com/mongodb/grip/recovery"
 	"github.com/pkg/errors"
 )
 
+// RetryableQueueOptions represent common options to configure an
+// amboy.RetryableQueue.
+type RetryableQueueOptions struct {
+	// Disabled is a function that determines whether or not retrying is
+	// disabled. If it returns true, jobs will not automatically retry. This
+	// allows retrying jobs to be toggled dynamically during runtime. By
+	// default, it will never be disabled. This setting always takes precedence
+	// over RetryHandler.Disabled.
+	Disabled func() bool
+	// RetryHandler are options to configure how retryable jobs are handled.
+	RetryHandler amboy.RetryHandlerOptions
+	// StaleRetryingMonitorInterval is how often a queue periodically checks for
+	// stale retrying jobs.
+	StaleRetryingMonitorInterval time.Duration
+}
+
+func (opts *RetryableQueueOptions) Validate() error {
+	catcher := grip.NewBasicCatcher()
+	catcher.NewWhen(opts.StaleRetryingMonitorInterval < 0, "stale retrying check frequency cannot be negative")
+	if opts.StaleRetryingMonitorInterval == 0 {
+		opts.StaleRetryingMonitorInterval = defaultStaleRetryingMonitorInterval
+	}
+	if opts.Disabled == nil {
+		opts.Disabled = func() bool { return false }
+	}
+	opts.RetryHandler.Disabled = opts.Disabled
```

Comment:
What is the point of having a `Disabled` field in `RetryHandlerOptions` if it always gets overwritten? Should this line only occur if `opts.Disabled != nil`?

Reply:
Also, why does there need to be two `Disabled` fields in the first place? I am a bit confused on how this function would work. Could you give an example?

Reply:
1. Because you can construct a retry handler without this struct and disable it. Disabled will never be nil. This is to handle queue initialization which includes retryable queue options, in which case the retry handler is initialized by the queue constructor. We need the separate setting because monitoring for stale retrying jobs is done at the queue level, not the retry handler level (i.e. `monitorStaleRetryingJobs()`).
2. We need two levels of disabling to deal with retry handler-level disabling and . `opts.Disabled` controls disabling at both the queue level (i.e. monitor stale retrying jobs) and the retry handler level (which is basically just a component of the queue).
3. It has to be a function because it's going to be based on a DB service flag in Evergreen/Cedar that can be toggled at runtime.